SpriteSheetPacker/MainWindow.cpp

```diff
@@ -606,27 +606,24 @@ void MainWindow::on_actionPublish_triggered() {
         }
     }
 
-    publisher->publish(ui->dataFormatComboBox->currentText(), ui->optModeComboBox->currentText(), ui->optLevelSlider->value());
+    publisher->publish(ui->dataFormatComboBox->currentText(), ui->optModeComboBox->currentText(), ui->optLevelSlider->value(), true, publishStatusDialog);
+
+        if (ui->optModeComboBox->currentText() == "None") {
+            publishStatusDialog->log(QString("Publishing is finished."), Qt::blue);
+            publishStatusDialog->complete();
+            publishStatusDialog->progressBar()->setValue(3);
 
-    if (ui->optModeComboBox->currentText() == "None") {
-        delete publisher;
-    } else {
-        // TODO: you need to wait previous image optimization if they are the same
-        QObject::connect(publisher, &PublishSpriteSheet::onCompleted, [this, publisher] () {
-            // TODO: it would be good to show here for information about the optimization
-            QMessageBox::information(this, "PNG Optimization", "PNG Optimization: complete.");
             delete publisher;
-        });
-    }
 
-    publishStatusDialog.log(QString("Publishing is finished."), Qt::blue);
-    if (!publishStatusDialog.complete()) {
-        // This loop will wait for the window is finished
-        QEventLoop loop;
-        connect(&publishStatusDialog, SIGNAL(finished(int)), &loop, SLOT(quit()));
-        loop.exec();
-        qDebug() << "Finish publish";
-    }
+        } else {
+            QObject::connect(publisher, &PublishSpriteSheet::onCompleted, [publishStatusDialog, publisher] () {
+                publishStatusDialog->log(QString("Publishing is finished."), Qt::blue);
+                publishStatusDialog->complete();
```

Comment:
Correct, the only button to close the dialog is the complete button that can only be pressed when optimization finished. It's probably a good idea to add a cancel button later.